### ISSUE TITLE
Add "Show Query" button next to "Filter" button 

### ISSUE THREAD

**Issue by samuelbradshaw** [Sequel-Ace/Sequel-Ace#290]
It would be great if there were a "Show Query" button next to the "Filter" button in the content filters area (see arrow in screenshot). I know that queries are available in the Console after they're executed, but having an option here would allow you to see a preview of the filter query you're about to run, before you run it.

I would expect that clicking "Show Query" would add a row with a read-only text area (similar to the text area in the "Advanced Filter Content" panel), immediately above the row with the Show Query / Filter buttons. The "Show Query" button would change to "Hide Query" while the query row is visible.

<img width="1130" alt="Screen Shot 2020-08-10 at 11 50 05 AM" src="https://user-images.githubusercontent.com/7323052/89814050-caabe500-daff-11ea-911c-d5294c917bad.png">

**Comment by jamesstout** [comment]
Sort of got something working ... needs work though.

![2020-08-11 06 02 07](https://user-images.githubusercontent.com/1576190/89836022-7acd2c80-db98-11ea-86c4-9a4d37e6fdb9.gif)

**Comment by Jason-Morcos** [comment]
Perhaps instead of an actual show query text box on the main filter bar, it would make more sense to have a "preview query" button which launched an alert view (like the modal style alert view) with the query text and buttons to copy the SQL and execute the previewed filter string? Seems like a bigger text view would also make it easier to see the actual full SQL statement.

**Comment by samuelbradshaw** [comment]
Nice!

I noticed in your screen grab, you're executing the query before pressing Show Query – are you able to have the query update as you're adding and removing filters, similar to how the query shows and live-updates in the "Advanced Filter Content" panel? Or do you have it set up such that you have to execute the filter first, before you can see the query?

Re: text box vs. alert – I like having the query visible right there, without having to leave the current context, and I think it could be helpful for users who are learning SQL (if they've toggled "Show Query" on). Though if the query were more than one line, I'd want the query to wrap and the box to expand downward. @jamesstout, will it do that with what you have so far? @Jason-Morcos, assuming the inline text box were able to expand downward, what benefits do you see from having the query open in an alert/modal?

**Comment by Jason-Morcos** [comment]
> @Jason-Morcos, assuming the inline text box were able to expand downward, what other benefits do you see from having the query open in an alert/modal instead of being inline?

The big benefit I see is the ability to add additional capabilities like "copy filter query" or "export query as .SQL". It seems like adding an expanding textbook will further crowd the main table view. I think I would use the show query text feature as well, but I would use it to copy a query string to use somewhere else (in a project I'm working on or perhaps just to modify in the query editor manually - but I would have to toggle the query view visible, copy it over, then toggle invisible afterwards which seems annoying.
Additionally, real-time updating _sounds_ cool but that means another process running every time you type a character which seems to me to be distracting, kinda unnecessary, and potentially another source of lag.

**Comment by samuelbradshaw** [comment]
> toggle the query view visible, copy it over, then toggle invisible afterwards which seems annoying.
This is the exact annoyance I would have if it were an alert – I would have to click to open the alert, then copy, then close the alert. But if it's inline, it would just be there, and I could copy/paste it without toggling anything. :)

I don't think I would use an option to export a single query as SQL; mainly I'd be viewing and/or copying. I don't think a button for copying would provide much value either, as it's just as easy to Edit > Copy, right-click Copy, or ⌘C. If this feature ends up being an alert, it wouldn't hurt to have those buttons, but I don't think see those as a strong enough argument to take the user to another view.

I agree that making sure it doesn't lag is important. If needed, maybe it would only update when you pause for a couple of seconds, or if the box you're typing in loses focus. And of course it wouldn't update if Show Query is toggled off.

Being a project owner, you have the tie-breaking vote, but I'm curious if anyone else following this thread has an opinion.

**Comment by Jason-Morcos** [comment]
For this issue, I'll definitely yield to @jamesstout's wisdom haha. He's writing the code, whichever looks/feels/seems better to him is how we should move forward!

**Comment by jamesstout** [comment]
> Nice!
> I noticed in your screen grab, you're executing the query before pressing Show Query – are you able to have the query update as you're adding and removing filters, similar to how the query shows and live-updates in the "Advanced Filter Content" panel? 

Maybe ... I think the keys are captured already.

> Or do you have it set up such that you have to execute the filter first, before you can see the query? 

Can't remember .. I'll have to check the code again, but I think you have to filter first.

>  @jamesstout, is the text box able to grow downward with a long query? 

Not at the moment, it was a quick POC, I didn't add all the constraints etc.

**Comment by jamesstout** [comment]
> For this issue, I'll definitely yield to @jamesstout's wisdom haha. He's writing the code, whichever looks/feels/seems better to James is how we should move forward!

Wisdom 🧐 ... I'm an iOS developer helping out where I can :) 

I'll play around with it this weekend.

**Comment by jamesstout** [comment]
> Maybe ... I think the keys are captured already.

No, they are captured on the custom query view, but not from the filter

> 
> > Or do you have it set up such that you have to execute the filter first, before you can see the query?
> 
> Can't remember .. I'll have to check the code again, but I think you have to filter first.

Filter first.

**Comment by Robert-Saiter** [comment]
I like this idea... it could even be a label that only gets updated when the filter is modified.

**Comment by crazytonyi** [comment]
> The big benefit I see is the ability to add additional capabilities like "copy filter query" or "export query as .SQL". 

I was looking for Feature Requests for "Copy Filter Query" when I found this. Spot on. I want the ability to copy the filter query and paste it over in the Query window so I can refine it more (add GROUP BYs and maybe move columns around, etc). Export Query as .sql is a *really* great idea as well.

**Comment by Jason-Morcos** [comment]
@jamesstout Looking back through it looks like you had done some stuff on this issue - would it be something you'd want to merge in? Additionally, maybe instead of showing the query in a little box, what if "show query" put the query into the query editor and focused it? Less UI changes. Also would be fine, I think, if it was just a "copy query" button and the user could do with it what they please (paste into query editor or into another app or anything).

**Comment by samuelbradshaw** [comment]
A "Copy Query" button would be a good compromise for simplicity. I think that putting the query into the regular query editor automatically and bringing it to focus, when pressing a "Show Query" button, could be unexpected and potentially obtrusive from a user perspective, since it would 1) navigate you away from your current context, and 2) overwrite or clutter potentially unrelated tasks you're working on in the query editor.

**Comment by jacobhtye** [comment]
Did anything ever happen with this?

**Comment by Jason-Morcos** [comment]
No, but PR's are always welcome :)

**Comment by JonLaliberte** [comment]
This would be a great feature to add. I was just trying to find a way to copy the current filter query, thinking it would be one of the options in Edit. Previewing/Showing the query would be nice, but the Copy command would be a big time saver.
(Understood PRs are welcome, but I'm but a lowly user these days, having not fired up Xcode in ages :( )

**Comment by Jason-Morcos** [comment]
> This would be a great feature to add. I was just trying to find a way to copy the current filter query, thinking it would be one of the options in Edit. Previewing/Showing the query would be nice, but the Copy command would be a big time saver.
> (Understood PRs are welcome, but I'm but a lowly user these days, having not fired up Xcode in ages :( )

Don't forget that you can open the Console window and copy the filters query that was run right from there :)

**Comment by mjaggard** [comment]
I came here to make a very similar request. I was hoping for a button to "Show SQL in Query view" which would take the current view of data, based on the table, order and filter, and add it to a new line on the Query tab then show that tab. I can see that doing that before running the query might be useful but slightly edge-case.

**Comment by wojtiku** [comment]
Ability to somehow get the currently executed query would be great - it would serve as a great starting point to make refinements, extra joins etc. and would make me faster when using the app :)

**Comment by lxmorj** [comment]
Agreed, would love to see this.

**Comment by AdrianHL** [comment]
Would be great to see this feature in a future release. Thanks :)
